Sming/SmingCore/Wire.h

```diff
@@ -1,60 +1,91 @@
 
-class SoftI2cMaster;
+  Modified 2012 by Todd Krein (todd@krein.org) to implement repeated starts
+  Modified December 2014 by Ivan Grokhotkov (ivan@esp8266.com) - esp8266 support
+  Modified April 2015 by Hrsto Gochkov (ficeto@ficeto.com) - alternative esp8266 support
+*/
+
+#ifndef TwoWire_h
+#define TwoWire_h
+
+#include <inttypes.h>
+#include "Stream.h"
+
+#define BUFFER_LENGTH 32
 
 class TwoWire : public Stream
 {
-public:
-	TwoWire(int pinSCL, int pinSDA);
-	virtual ~TwoWire();
-
-	void pins(int pinSCL, int pinSDA); // Can be called only before begin()
-	void begin();
-
-	void beginTransmission(uint8_t address);
-	uint8_t endTransmission(bool sendStop = true);
-	uint8_t requestFrom(int address, int quantity, bool sendStop = true);
-
-	virtual int available();
-	virtual int read();
-	virtual int peek();
-	virtual void flush();
-	size_t write(uint8_t data);
-	size_t write(const uint8_t *data, size_t quantity);
-
-protected:
-	uint8_t pushData();
-
-private:
-	SoftI2cMaster* master;
-	int SCL;
-	int SDA;
-	int targetAddress;
-
-	uint8_t txBuf[BUFFER_LENGTH];
-	int txLen;
-
-	uint8_t rxBuf[BUFFER_LENGTH];
-	int rxLen;
-	int rxPos;
+  private:
+    static uint8_t rxBuffer[];
+    static uint8_t rxBufferIndex;
+    static uint8_t rxBufferLength;
+
+    static uint8_t txAddress;
+    static uint8_t txBuffer[];
+    static uint8_t txBufferIndex;
+    static uint8_t txBufferLength;
+
+    static uint8_t transmitting;
+    static void (*user_onRequest)(void);
+    static void (*user_onReceive)(int);
+    static void onRequestService(void);
+    static void onReceiveService(uint8_t*, int);
+  public:
+    TwoWire();
+    void begin(int scl, int sda);
+    void pins(int scl, int sda);
+    void begin();
+    void begin(uint8_t);
+    void begin(int);
+    void setClock(uint32_t);
+    void setClockStretchLimit(uint32_t);
+    void beginTransmission(uint8_t);
+    void beginTransmission(int);
+    uint8_t endTransmission(void);
+    uint8_t endTransmission(uint8_t);
+    size_t requestFrom(uint8_t address, size_t size, bool sendStop);
+	uint8_t status();
+
+    uint8_t requestFrom(uint8_t, uint8_t);
+    uint8_t requestFrom(uint8_t, uint8_t, uint8_t);
+    uint8_t requestFrom(int, int);
+    uint8_t requestFrom(int, int, int);
+    
+    virtual size_t write(uint8_t);
+    virtual size_t write(const uint8_t *, size_t);
+    virtual int available(void);
+    virtual int read(void);
+    virtual int peek(void);
+    virtual void flush(void);
+    void onReceive( void (*)(int) );
+    void onRequest( void (*)(void) );
+
+    inline size_t write(unsigned long n) { return write((uint8_t)n); }
+    inline size_t write(long n) { return write((uint8_t)n); }
+    inline size_t write(unsigned int n) { return write((uint8_t)n); }
+    inline size_t write(int n) { return write((uint8_t)n); }
+    using Print::write;
 };
 
+#if !defined(NO_GLOBAL_INSTANCES) && !defined(NO_GLOBAL_TWOWIRE)
 extern TwoWire Wire;
```

Comment:
It would be better, memory wise, to avoid creating global objects.  Therefore I would prefere the global instance to be created only if requested (in the code above it is created by default).